functions/functions.go

```diff
@@ -1627,3 +1627,144 @@ func (Wood) Minima() []Minimum {
+//  Decrease. ACM Transactions on Mathematical Software 20(3) (1994), 286–307
+// The function is concave to the right of the minimizer that is located at x=sqrt(2).
+type ConcaveRight struct{}
+
+func (ConcaveRight) Func(x []float64) float64 {
+	if len(x) != 1 {
+		panic("dimension of the problem must be 1")
+	}
+
+	return -x[0] / (x[0]*x[0] + 2)
+}
+
+func (ConcaveRight) Grad(grad, x []float64) {
+	if len(x) != 1 {
+		panic("dimension of the problem must be 1")
+	}
+	if len(x) != len(grad) {
+		panic("incorrect size of the gradient")
+	}
+
+	xSqr := x[0] * x[0]
+	grad[0] = (xSqr - 2) / (xSqr + 2) / (xSqr + 2)
+}
+
+// ConcaveLeft implements the univariate function labeled as (5.2) in
+//  More, J.J., and Thuente, D.J.: Line Search Algorithms with Guaranteed Sufficient
+//  Decrease. ACM Transactions on Mathematical Software 20(3) (1994), 286–307
+// The function is concave to the left of the minimizer that is located at x=399/250=1.596.
+type ConcaveLeft struct{}
+
+func (ConcaveLeft) Func(x []float64) float64 {
+	if len(x) != 1 {
+		panic("dimension of the problem must be 1")
+	}
+
+	return math.Pow(x[0]+0.004, 4) * (x[0] - 1.996)
+}
+
+func (ConcaveLeft) Grad(grad, x []float64) {
+	if len(x) != 1 {
+		panic("dimension of the problem must be 1")
+	}
+	if len(x) != len(grad) {
+		panic("incorrect size of the gradient")
+	}
+
+	grad[0] = math.Pow(x[0]+0.004, 3) * (5*x[0] - 7.98)
+}
+
+// Plassmann implements the univariate function labeled as (5.3) in
+//  More, J.J., and Thuente, D.J.: Line Search Algorithms with Guaranteed Sufficient
+//  Decrease. ACM Transactions on Mathematical Software 20(3) (1994), 286–307
+type Plassmann struct {
+	L    float64 // Number of oscillations for |x-1| ≥ Beta.
+	Beta float64 // Size of the derivative at zero, f'(0) = -Beta.
+}
+
+func (f Plassmann) Func(x []float64) float64 {
+	if len(x) != 1 {
+		panic("dimension of the problem must be 1")
+	}
+
+	a := x[0]
+	b := f.Beta
+	l := f.L
+	r := 2 * (1 - b) / l / math.Pi * math.Sin(l*math.Pi/2*a)
+	switch {
+	case a <= 1-b:
+		r += 1 - a
+	case 1-b < a && a <= 1+b:
+		r += 0.5 * ((a-1)*(a-1)/b + b)
+	default: // a > 1+b
+		r += a - 1
+	}
+	return r
+}
+
+func (f Plassmann) Grad(grad, x []float64) {
+	if len(x) != 1 {
+		panic("dimension of the problem must be 1")
+	}
+	if len(x) != len(grad) {
+		panic("incorrect size of the gradient")
+	}
+
+	a := x[0]
+	b := f.Beta
+	l := f.L
+	grad[0] = (1 - b) * math.Cos(l*math.Pi/2*a)
+	switch {
+	case a <= 1-b:
+		grad[0] -= 1
+	case 1-b < a && a <= 1+b:
+		grad[0] += (a - 1) / b
+	default: // a > 1+b
+		grad[0] += 1
+	}
+}
+
+// YanaiOzawaKaneko implements the univariate function labeled as (5.4) in
+//  More, J.J., and Thuente, D.J.: Line Search Algorithms with Guaranteed Sufficient
+//  Decrease. ACM Transactions on Mathematical Software 20(3) (1994), 286–307
+// See also
+//  Yanai, H., Ozawa, M., and Kaneko, S.: Interpolation methods in one dimensional
+//  optimization. Computing 27 (1981), 155–163
```

Comment:
Very nice description, thanks! I wanted to mention the absolute value function but couldn't formulate it in a not-too-mathematical way. Done.